src/session/session_api.c

```diff
@@ -1423,6 +1435,37 @@ __open_session(WT_CONNECTION_IMPL *conn,
 	session_ret->id = i;
 	session_ret->iface = stds;
 	session_ret->iface.connection = &conn->iface;
+	/*
+	 * Disable some methods if this is a read-only connection.  Group
+	 * them by call signature.
+	 */
+	if (F_ISSET(conn, WT_CONN_READONLY)) {
+		wt_session = &session_ret->iface;
+		wt_session->checkpoint =
+		    (int (*)(WT_SESSION *, const char *))__wt_session_notsup;
+		wt_session->compact =
+		    (int (*)(WT_SESSION *, const char *, const char *))__wt_session_notsup;
```

Comment:
Hmm, that article also says this - which sounds exactly like what we're doing unless I'm reading it wrong:

Calling a function pointer boils down to pushing the arguments on the stack and doing an indirect jump to the function pointer target, and there's obviously no notion of types at the machine code level.

Things you definitely can't do:

    Cast between function pointers of different calling conventions. You will mess up the stack and at best, crash, at worst, succeed silently with a huge gaping security hole.